llama_index/vector_stores/pinecone.py

```diff
@@ -210,21 +213,33 @@ def __init__(
     ) -> None:
         """Initialize params."""
         try:
-            import pinecone
+            # Dynamic import of pinecone module, needed to simulate different versions of client in tests
+            pinecone = __import__("pinecone")
         except ImportError:
             raise ImportError(import_err_msg)
 
         if pinecone_index is not None:
```

Comment:
@aulorbe As you pointed out in direct messages, now that I look at this again, it does appear that the logic is off here: 

Currently, we're only running the version checks if the user supplies an existing Pinecone index name, but we instead want to first always run the version checking logic, and then set the internal reference to the Pinecone client to either the client the user passed in, or the default that is initialized within the `init` method itself. 

So, something like: 

```python
# Check Pinecone version 
if version.parse(pinecone_client_version) >= version.parse("3.0.0"):
   # Init new client
else:
   # Init old client
   
# Then either use the initialized client or 
# the passed in pinecone_index 
if pinecone_index:
   self._pinecone_index = pinecone_index  
else:
   self._pinecone_index = # the initialized client above
```

Great catch!

Reply:
Yes, will try to refactor today